src/watchers/indexable-post.php

```diff
@@ -0,0 +1,322 @@
+			return;
+		}
+	}
+
+	/**
+	 * Saves post meta.
+	 *
+	 * @param int $post_id Post ID.
+	 *
+	 * @return void
+	 */
+	public function save_meta( $post_id ) {
+		if ( ! $this->is_post_indexable( $post_id ) ) {
+			return;
+		}
+
+		try {
+			$indexable = $this->get_indexable( $post_id );
+		} catch ( No_Indexable_Found $exception ) {
+			return;
+		}
+
+		$indexable->permalink = $this->get_permalink( $post_id );
+
+		foreach ( $this->get_meta_lookup() as $meta_key => $indexable_key ) {
+			$indexable->{$indexable_key} = $this->get_meta_value( $meta_key, $post_id );
+		}
+
+		$indexable->primary_focus_keyword_score = $this->get_keyword_score(
+			$this->get_meta_value( 'focuskw', $post_id ),
+			$this->get_meta_value( 'linkdex', $post_id )
+		);
+
+		$indexable->is_robots_noindex = $this->get_robots_noindex( $this->get_meta_value( 'meta-robots-noindex', $post_id ) );
+
+		// Set additional meta-robots values.
+		$nonidex_advanced = $this->get_meta_value( 'meta-robots-adv', $post_id );
+		$meta_robots      = explode( ',', $nonidex_advanced );
+		foreach ( $this->get_robots_options() as $meta_robots_option ) {
+			$indexable->{'is_robots_' . $meta_robots_option} = in_array( $meta_robots_option, $meta_robots, true ) ? 1 : null;
+		}
+
+		$indexable = $this->set_link_count( $post_id, $indexable );
+
+		$indexable->save();
+
+		if ( ! empty( $indexable->id ) ) {
+			$this->save_social_meta( $indexable, $post_id );
+		}
+	}
+
+	/**
+	 * Saves the social meta data.
+	 *
+	 * @param Indexable $indexable The indexable to save the id for.
+	 * @param int       $post_id   Post ID.
+	 *
+	 * @codeCoverageIgnore
+	 */
+	protected function save_social_meta( $indexable, $post_id ) {
+		$indexable_post_meta = new Indexable_Meta( $indexable->id );
+
+		foreach ( $this->social_meta_lookup() as $meta_key => $indexable_key ) {
+			$indexable_post_meta->set_meta( $indexable_key, $this->get_meta_value( $meta_key, $post_id ) );
+		}
+	}
+
+	/**
+	 * Determines if the post can be indexed
+	 *
+	 * @codeCoverageIgnore
+	 *
+	 * @param int $post_id Post ID to check.
+	 *
+	 * @return bool True if the post can be indexed.
+	 */
+	protected function is_post_indexable( $post_id ) {
+		if ( \wp_is_post_revision( $post_id ) ) {
+			return false;
+		}
+
+		if ( \wp_is_post_autosave( $post_id ) ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Fetches the indexable for a post.
+	 *
+	 * @param int  $post_id     Post to fetch indexable for.
+	 * @param bool $auto_create Optional. Create the indexable if it does not exist.
+	 *
+	 * @return Indexable
+	 *
+	 * @throws No_Indexable_Found Exception when no Indexable entry could be found.
+	 */
+	protected function get_indexable( $post_id, $auto_create = true ) {
+		$post_type = $this->get_post_type( $post_id );
+		$indexable = Yoast_Model::of_type( 'Indexable' )
+								->where( 'object_id', $post_id )
+								->where( 'object_type', 'post' )
+								->where( 'object_sub_type', $post_type )
```

Comment:
I'm not sure we'd want to query on `post_type` here. Using plugins like Post Type Switcher the post type of a post may change, in which case this would find nothing and a new indexable would be created.

The combination of `object_id` and `object_type` is what's identifiable and unique, I don't think `object_sub_type` should be treated the same.

I would set `object_sub_type` in the `set_meta` function the same as `permalink`.